### PR TITLE
AppFooter Implementation

### PR DESCRIPTION
**Pull request by bluekurinji**

### CODE REVIEW

**Comment by chrisolsen** [comment]
Can you change the commit message to follow the guidelines ex. `feat(#999): app-footer component`

File: libs/react-components/src/lib/app-footer/meta-link.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace JSX {
+    interface IntrinsicElements {
+      'goa-meta-link': WCProps & React.HTMLAttributes<HTMLElement>
+    }
+  }
+}
+
+interface Props {
+  footerid?: string;
```
Comment (chrisolsen) [inline libs/react-components/src/lib/app-footer/meta-link.tsx:19]: For the React props use the camelCase style

File: libs/react-components/src/lib/app-footer/navigation-link.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace JSX {
+    interface IntrinsicElements {
+      'goa-navigation-link': WCProps & React.HTMLAttributes<HTMLElement>
+    }
+  }
+}
+
+interface Props {
+  footerid?: string;
```
Comment (chrisolsen) [inline libs/react-components/src/lib/app-footer/navigation-link.tsx:20]: Use the camelCase here as well

File: libs/web-components/demo/src/App.svelte [hunk 3]
```diff
@@ -14,17 +14,11 @@
...
-    </div>
-    Lorem ipsum dolor sit amet consectetur adipisicing elit. Consectetur optio, eius expedita,
-    quasi vero maxime, laudantium voluptates eaque illum officiis hic! Commodi, asperiores
-    officiis necessitatibus itaque similique magnam nisi voluptate.
-  </goa-hero-banner>
+  <center>
+    <br>
+    <h1> content here </h1>
+    <br>
+  </center>
```
Comment (chrisolsen) [inline libs/web-components/demo/src/App.svelte:21]: Probably want to revert this change

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import GoAAppFooter from './AppFooterWrapper.test.svelte'
+
+function createElement(props = {}) {
+  return render(GoAAppFooter, { ...props });
+}
+
+describe('GoAAppFooter Component', () => {
+
+  it("should render", async () => {
+    const el = await createElement();
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.spec.ts:12]: I don't think `await` is required here

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+      const urls: string[] = [];
+      const title: string[] = [];
+      metaLinks.forEach(function(metaLink) {
+        const linkElement = (metaLink as HTMLAnchorElement);
+        urls.push(linkElement.href);
+        title.push(linkElement.text);
+      });
+
+      expect(["http://localhost/A", "http://localhost/B", "http://localhost/C", "http://localhost/D"]).toEqual(urls.sort());
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.spec.ts:39]: The params passed into the `expect()` function need to be flipped with the ones in the `toEqual`. The test passes now, but if it later fails for whatever reason the error message will be confusing; it was confusing when I was looking at the tests this morning.

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+  }
+
+  onMount(async () => {
+    ctx = createContext(id);
+    ctx.subscribe((state) => {
+      switch (state?.type) {
+        case META_LINK: {
+          const message = state as MetaLinkRegisterMessage;
+          metaLinks = [...metaLinks, {title: message.title, url: message.url}];
+          console.log(metaLinks);
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:62]: remove

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+    box-sizing: border-box;
+    font-family: var(--font-family);
+  }
+
+  .footer {
+    max-width: 960px;
+  }
+
+  .brand {
+    height: 16px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:156]: Replace all `px` values with the `rem` equivalent

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+
+<!-- Style -->
+<style>
+  :host {
+    box-sizing: border-box;
+    font-family: var(--font-family);
+  }
+
+  .footer {
+    max-width: 960px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:152]: I don't think we should be defining the max width there. It should probably be controlled by it's parent container element.

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+  .meta-link {
+    margin-top:56px;
+    margin-right: 28px;
+    color: var(--goa-color-text);
+    font-size: var(--fs-base);
+  }
+
+  .logo {
+    height:41px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:185]: The height and width are already set within the svg element; is there a reason they are redefined here?

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+  .navigation-links, .meta-links, .navigation-section {
+    text-align: start;
+    display: flex;
+    flex-flow: wrap;
+  }
+
+  .navigation-section {
+    flex-direction: column;
+    flex: 1 1 0;
+    min-width: 212x;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:202]: This is an odd number. Is this defined in the UI specs?

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+  }
+
+  .default-footer .logo-and-copyright {
+    display: flex;
+    align-items: center;
+    justify-content: space-between;
+  }
+
+  .meta-links-only-footer {
+    padding-top: 19px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:233]: This looks like a value that was tweaked until it looked right :) If this was in the design specs you will need to talk to them again to get a number that is consistent with their design guidelines.

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+    font-size: var(--fs-xl);
+    line-height: var(--lh-lg);
+    font-weight: var(--fw-regular);
+    margin-bottom: 28px;
+    margin-top: 28px;
+  }
+
+  .navigation-section-name-divider {
+    width: 75%;
+    margin: 0px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:215]: Not sure if setting the margin to zero is required, but if it is just set it to `0` to reduce the css warnings.

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+  .navigation-links-only-footer,
+  .meta-and-navigation-links-only-footer,
+  .meta-and-navigation-sections-only-footer,
+  .navigation-sections-only-footer  {
+    padding-top: 28px;
+    padding-bottom:56px;
+  }
+
+  .meta-links-only-footer .meta-link {
+    margin-top: 37.5px;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:246]: 37.5px...

File: libs/web-components/src/components/app-footer/MetaLink.svelte [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+</script>
+
+<!-- HTML -->
+<data
+  bind:this={metaLink}
+  data-url={url}
+  data-title={title}
+>
+</data>
+
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/MetaLink.svelte:35]: What is the reason for this data element?

File: libs/web-components/src/components/app-footer/NavigationLink.svelte [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+<!-- HTML -->
+<data
+  bind:this={navigationLink}
+  data-url={url}
+  data-title={title}
+  data-section={section}
+>
+</data>
+
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/NavigationLink.svelte:38]: I don't think these data elements are required.

File: libs/web-components/src/components/app-footer/types.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import type { Message } from "../../common/context-store"
+
+export const META_LINK = "meta-link";
+export const NAVIGATION_LINK = "navigation-link";
+
+export interface MetaLinkRegisterMessage extends Message {
+  type: "meta-link" | "navigation-link";
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/types.ts:7]: Why is this type set to meta-link or navigation-link, since it is a MetaLinkRegisterMessage I would think that it should be just 'meta-link'. The same is true for the NavigationLInkRegisterMessage below, except flipped.

File: libs/web-components/src/components/app-footer/navigationSection.ts [hunk 17]
```diff
@@ -0,0 +1,6 @@
+import type { Link } from "./link"
+
+export interface NavigationSection {
+  name: string | null;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/navigationSection.ts:4]: Can this be changed to `name?: string`?

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts (hunk 5 above)
Comment (chrisolsen) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:39]: The same is true for some of the other `expect` calls within other tests.

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    const el = await createElement({ navigationSections: navigationsections });
+
+    await waitFor(() => {
+
+      const sectionNames = el.container.querySelectorAll('.navigation-section-name');
+      const names: string[] = [];
+      sectionNames.forEach(function(sectionName) {
+        const element = (sectionName as HTMLElement);
+        names.push((element.firstChild as HTMLElement).textContent);
+      });
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.spec.ts:96]: This might be a little cleaner using `map` rather that a `forEach`, same for any other `forEach` calls.

Comment (chrisolsen) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:96]: But using the `expect(el).toContain(...)` might be even more clear.

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+              {/each}
+            </div>
+          {/each}
+        {:else if navigationLinks.length }
+          {#each navigationLinks as navigationlink (navigationlink.title) }
+            <a href={navigationlink.url} class="navigation-link">{navigationlink.title}</a>
+          {/each}
+        {/if}
+      </div>
+      <hr class="navigation-links-divider" />
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:114]: Missing indent

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+      {/if}
+
+      <div class:meta-links-logo-and-copyright="{metaLinks.length}">
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:117]: The quotes might result in a `0` length resolving to `true`.

Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte:116]: Svelte nicely handles it. For consistency, I have removed it

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 8 above)
Comment (ravi-dio) [reply libs/web-components/src/components/app-footer/AppFooter.svelte]: I see max-width of 60rem - is it OK to have a max width here based on Chris's comment? Also, shouldn't footer span the entire width of the page?

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts (hunk 18 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:95]: This is unit test with 3 lines of code. So, I  prefer to leave this as is.

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 11 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte]: The logo's dimension has some impact on rest of the elements spacing. The look and feel is made to meet design specifications.

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 10 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte]: This will be discussed with designer during designer sign-off. The discussion is in progress

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 8 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte]: It is in correct place. The footer content has to be of max width 960px or 60rem. The parent allows room for padding when the view port is less than 960px as per the specification. Also, when the resolution is more than 960px, it keeps the spacing around 960px content.

File: libs/web-components/demo/src/App.svelte [hunk 21]
```diff
@@ -14,18 +14,6 @@
...
-    backgroundurl="https://picsum.photos/id/1076/600/400"
-  >
-    <div slot="actions">
-      <goa-button>Button</goa-button>
-    </div>
-    Lorem ipsum dolor sit amet consectetur adipisicing elit. Consectetur optio, eius expedita,
-    quasi vero maxime, laudantium voluptates eaque illum officiis hic! Commodi, asperiores
-    officiis necessitatibus itaque similique magnam nisi voluptate.
-  </goa-hero-banner>
-
```
Comment (chrisolsen) [inline libs/web-components/demo/src/App.svelte:28]: Can you add this back in?

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+function createElement(props = {}) {
+  return render(GoAAppFooter, { ...props });
+}
+
+describe('GoAAppFooter Component', () => {
+
+  it("should render", async () => {
+    const el = createElement();
+    const appFooter = el.container.querySelectorAll('.app-footer-container');
+    expect(appFooter).toBeTruthy();
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.spec.ts:14]: querySelectorAll will return a list and an empty list is still truthy. So you should instead use `querySelector`

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+      const navigationLinks = el.container.querySelectorAll('.navigation-link');
+      const urls: string[] = [];
+      const title: string[] = [];
+      navigationLinks.forEach(function(navigationLink) {
+        const linkElement = (navigationLink as HTMLAnchorElement);
+        urls.push(linkElement.href);
+        title.push(linkElement.text);
+      });
+
+      expect(urls.sort()).toEqual(["http://localhost/A", "http://localhost/B", "http://localhost/C", "http://localhost/D"]);
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.spec.ts:69]: Where is the protocol and domain coming from?

File: libs/web-components/src/components/app-footer/AppFooter.svelte [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+  .meta-link {
+    margin-top: 3.5em;
+    margin-right: 1.75rem;
+    color: var(--goa-color-text);
+    font-size: var(--fs-base);
+  }
+
+  .logo {
+    height: 2.56rem;
+    width: 8.88rem;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/AppFooter.svelte:185]: - All these padding and margins is way to hacky. 
- Some `em` units are being used others use `rem`
- Breakpoints need to use px rather than rem

File: libs/web-components/src/components/app-footer/MetaLink.svelte [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+    ctx.notify({
+      type: META_LINK,
+      url: url,
+      title: title
+    });
+  });
+
+  export let footerid: string = "goa-app-footer-id";
+  export let title: string = "";
+  export let url: string = "";
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/MetaLink.svelte:22]: Move these to the top under the imports

File: libs/web-components/src/components/app-footer/NavigationLink.svelte [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+      section: section,
+    });
+  });
+
+  let navigationLink;
+
+  export let footerid: string = "goa-app-footer-id";
+  export let title: string = "";
+  export let url: string = "";
+  export let section: string = "";
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/NavigationLink.svelte:27]: same with these

File: libs/web-components/src/components/app-footer/NavigationLink.svelte [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+    ctx.notify({
+      type: NAVIGATION_LINK,
+      url: url,
+      title: title,
+      section: section,
+    });
+  });
+
+  let navigationLink;
```
Comment (chrisolsen) [inline libs/web-components/src/components/app-footer/NavigationLink.svelte:22]: what is this variable for?

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 24 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte:185]: The dimensions are arrived at support the design.  I could not locate any em units usage.

Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte:185]: Regarding px used for break points in media queries, please find the earlier discussion

Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.svelte:185]: We need to have a sperate discussion about which unit to use where. I dont want to hold this pull request until then

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts (hunk 23 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:69]: The protocol and domain comes from the test framework. I do not have much information at this time. This should not keep the pull request from moving forward

File: libs/web-components/demo/src/App.svelte (hunk 21 above)
Comment (bluekurinji) [reply libs/web-components/demo/src/App.svelte]: done

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts (hunk 22 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:14]: done

File: libs/web-components/src/components/app-footer/MetaLink.svelte (hunk 25 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/MetaLink.svelte]: done

File: libs/web-components/src/components/app-footer/NavigationLink.svelte (hunk 26 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/NavigationLink.svelte]: done

File: libs/web-components/src/components/app-footer/NavigationLink.svelte (hunk 27 above)
Comment (bluekurinji) [reply libs/web-components/src/components/app-footer/NavigationLink.svelte]: dead code. removed

File: libs/web-components/src/components/app-footer/AppFooter.spec.ts (hunk 23 above)
Comment (chrisolsen) [reply libs/web-components/src/components/app-footer/AppFooter.spec.ts:69]: It's not

File: libs/web-components/src/components/app-footer/AppFooter.svelte (hunk 24 above)
Comment (chrisolsen) [reply libs/web-components/src/components/app-footer/AppFooter.svelte:185]: Looks like I misunderstood your previous question, but as you stated we only used px values.

**Review by ravi-dio — APPROVED** [review]
